execUtil.go

```diff
@@ -1,62 +1,155 @@
-      err = cmd.Wait()
-      log.Printf("Command finished with error: %v", err)
+	var s StatusResponse
+	s.Task_id = task_id
 
-      // TODO mark cmd as finished in subroutine
-  }()
-  hostname, err := os.Hostname()
+	var command *CommandWrapper
+	var ok bool
 
-  // for now make this for unique task_id
-  task_id := hostname + "-" + strconv.Itoa(cmd.Process.Pid)
-  TaskIdCmdMap[task_id] = cmd
+	// validate task_id
+	if command, ok = taskIdCommandMap.Get(task_id); !ok {
+		s.Errors = []string{"invalid task_id"}
+		return s
+	}
 
-  log.Printf("RunCommand task_id=%s, pid=%d", task_id, TaskIdCmdMap[task_id].Process.Pid)
+	// TODO refactor this into un/finished process
+	s.Task_id = task_id
+	s.StartTime = new(time.Time)
+	*s.StartTime = command.StartTime
+	s.Finished = new(bool)
+	*s.Finished = command.GetFinished()
+	s.Errors = command.GetErrors()
 
-  return task_id
+	// cmd.Wait() has finished, append
+	if command.GetFinished() {
+		s.EndTime = new(time.Time)
+		*s.EndTime = command.EndTime
+		s.ExitCode = new(int)
+		*s.ExitCode = command.GetExitCode()
+	}
+
+	s.Output = command.StdoutBuff.Lines()
+
+	return s
 }
 
-// @param pid - stops process with this pid
-func StopProcess(task_id string) int {
+// RunCommand starts a process from command argument.
+func RunCommand(command string) StartResponse {
+	var s StartResponse
+
+	cmd := exec.Command(command)
+	cmd.SysProcAttr = &syscall.SysProcAttr{Setpgid: true}
+
+	outBuf := NewOutput()
+	cmd.Stdout = io.MultiWriter(os.Stdout, outBuf)
+
+	err := cmd.Start()
+
+	if err != nil {
+		s.Error = err.Error()
+		return s
+	}
+
+	pgid, err := syscall.Getpgid(cmd.Process.Pid)
+
+	if err != nil {
+		s.Error = err.Error()
+		return s
+	}
+
+	task_id := hostname + "-" + strconv.Itoa(pgid) // TODO handle if Process or Pid nil
+	s.Task_id = task_id
+
+	taskIdCommandMap.Put(task_id, NewCommandWrapper(cmd, outBuf))
+
+	// async subroutine
+	go func() {
+		// TODO add append error to CommandWrapper, impl accessors and setters
+		err = cmd.Wait()
+		cw, _ := taskIdCommandMap.Get(task_id)
+		if err != nil {
+			cw.AppendError(err.Error())
+		}
+		cw.SetExitCode(cmd.ProcessState.ExitCode())
+		cw.SetFinished(true)
+	}()
+
+	return s
+}
+
+// StopProcess is called to end a previously started process.
+func StopProcess(task_id string) StopResponse {
+
+	var s StopResponse
+	s.Task_id = task_id
+
+	var command *CommandWrapper
+	var ok bool
+
+	// if task_id invalid, return error.
+	if command, ok = taskIdCommandMap.Get(task_id); !ok {
+		s.Error = []string{"invalid task_id"}
+		return s
+	}
+
+	s.ExitCode = new(int)
+
+	if command.GetFinished() {
+		s.Error = []string{"process already finished."}
+		*s.ExitCode = command.GetExitCode()
+		return s
+	}
+
+	// send stop signal first
+	// TODO verify flat processes work with gpid as well
+
+	pgid, err := syscall.Getpgid(command.Command.Process.Pid)
+	err = syscall.Kill(-pgid, syscall.SIGINT)
+
+	if err != nil {
```

Comment:
nit: if this is permission issue - there's no point retrying with `SIGKILL`

Reply:
replacing with

`	if err != nil && err != syscall.EPERM {
`